### PR TITLE
feat: constrain union type `value` to actual allowed types instead of `any`

### LINKED ISSUE #758: Union types accept "any" value
When referencing the value of a generated union type, such as `k8s.Quantity`, the value is an `any` type. This can lead to unexpected output that does not meet the schemas being rendered at runtime, due to the lack of type checking.

For example, given a k8s.EnvVar like: 
```
const cpu = Quantity.fromNumber(2)
{
  name: "GOMAXPROCS",
  // EnvVar.value is of type 'string | undefined'
  // cpu.value is of type 'any', which fulfills 'undefined'
  value: cpu.value,
},
```

Because type checking is skipped for an `any`, and `cpu` was created from a number, when this value is rendered to JSON it is still a number, and thus the rendered JSON is invalid:
```
// Actual, invalid result:
{
  "name": "GOMAXPROCS",
  "value": 2
}
// Desired, valid result:
{
  "name": "GOMAXPROCS",
  "value": "2"
}
```

In order to provide useful type data, instead of `any`, I propose that the type of the union's value should be a union of the possible input types.

With this change, the example from above now will fail type checking, and provide a useful error:
```
// This code now fails to compile
const cpu = Quantity.fromNumber(2)
{
  name: "GOMAXPROCS",
  // EnvVar.value is of type 'string | undefined'
  // cpu.value is of type 'string | number', which cannot fulfill the expected type
  value: cpu.value,
},
```

```
// This code compiles and results in valid JSON that meets the schemas
const cpu = Quantity.fromNumber(2)
{
  name: "GOMAXPROCS",
  value: cpu.value.toString(),
},
```

### PR DESCRIPTION
**Pull request by heckler1**
Fixes #758

I believe this still retains compliance with the JSII spec, based on https://aws.github.io/jsii/specification/2-type-system/#type-unions: "...TypeScript models those cases using type unions, which are represented as `TypeA | TypeB`. The jsii type model supports those..."

Additionally, test renders of the k8s schemas via cdk8s-cli into python seemed to have the expected result, and tests all pass still.

### CODE REVIEW

File: src/type-generator.ts [hunk 1]
```diff
@@ -361,15 +361,17 @@ export class TypeGenerator {
       this.emitDescription(code, fqn, def.description);
 
       code.openBlock(`export class ${typeName}`);
+      const possibleTypes = [];
 
       for (const type of options) {
+        possibleTypes.push(type[0]+type.substr(1));
```
Comment (Chriscbr) [inline src/type-generator.ts:367]: Would this have the same effect?
```suggestion
        possibleTypes.push(type);
```

**Review by Chriscbr — COMMENTED** [review]
I think this change is ok since JSII should do the right thing and treat the union as "any" in languages like Java, but I'll ping @RomainMuller and @iliapolo to see if they can get a second look.

I'd also be curious if it's possible to generate a diff of how this changes the `imports/k8s.ts` generated by cdk8s -- maybe you could try sharing in a comment or gist if you have a chance? @heckler1

**Comment by heckler1** [comment]
This is the diff between a k8s.ts generated with `cdk8s import k8s@1.19.0 --language typescript`, vs the same command but with a cdk8s-cli that has a modified json2jsii package with this change.

```
➜ diff -U3 imports/k8s.ts imports/k8s-new-union.ts
--- imports/k8s.ts	2022-06-13 16:36:50.000000000 -0500
+++ imports/k8s-new-union.ts	2022-06-13 16:36:14.000000000 -0500
@@ -26149,7 +26149,7 @@
   public static fromNumber(value: number): Quantity {
     return new Quantity(value);
   }
-  private constructor(public readonly value: any) {}
+  private constructor(public readonly value: string | number) {}
 }

 /**
@@ -29619,7 +29619,7 @@
   public static fromNumber(value: number): IntOrString {
     return new IntOrString(value);
   }
-  private constructor(public readonly value: any) {}
+  private constructor(public readonly value: string | number) {}
 }

 /**
```

File: src/type-generator.ts (hunk 1 above)
Comment (heckler1) [reply src/type-generator.ts]: whoops, yes. Fixed, thanks!

**Comment by heckler1** [comment]
https://gist.github.com/heckler1/1f6781d626044da0e6c5a23155937c28

I also generated a test import for Java. I'm pretty unfamiliar with Java on the whole, but this looks right to me.

**Comment by Chriscbr** [comment]
Before adding this I think we need a unit test checking that this works in the case of non-primitive case. For example, suppose instead of the choices being `number | string` it was `TypeA | TypeB`.

**Comment by heckler1** [comment]
Hey there! I just got back around to looking at this again, and I realized, we don't support union types of non-primitives to begin with, so I think this should be fine with just the existing test for primitive unions.

https://github.com/cdklabs/json2jsii/blob/main/src/type-generator.ts#L8

**Comment by heckler1** [comment]
@Chriscbr hey hey, just bumping this, was wondering about your thoughts on my last comment?

**Review by Chriscbr — APPROVED** [review]

**Comment by heckler1** [comment]
Sorry to bump again but I think I need the build run approved as a first-time contributor to this repo?

**Comment by heckler1** [comment]
@Chriscbr just checking in on this, I think we just need a a build run done before we can merge

**Comment by Chriscbr** [comment]
@heckler1 unfortunately I don't have permission to merge this so someone else will need to review this. @iliapolo or @RomainMuller  perhaps?

**Review by Chriscbr — APPROVED** [review]

**Comment by heckler1** [comment]
Hi there! Just bumping this, was hoping we could get this one closed out

**Comment by heckler1** [comment]
@rix0rrr Is this something you might be able to take a look at?

**Review by iliapolo — COMMENTED** [review]
Nice, thanks! 

Really appreciated your thorough work and patience on this 👍

**Review by iliapolo — APPROVED** [review]

**Pull request merged**